peregrine/resources/submission/graphql/node.py

```diff
@@ -795,6 +797,67 @@ def _make_inner_meta_type():
     return gql_object
 
 
+class NodeCounter:
+    """A workaround for performance issue in PXD-2336.
+
+    With large dictionary of hundreds of nodes, constructing count queries can be a time
+    consuming task with SQLAlchemy. This workaround uses raw SQL and combines all simple
+    count queries into one to reduce round-trip time to database.
+    """
+    query_tpl = "(SELECT count(*) FROM %s WHERE _props->>'project_id' = :%s) AS %s"
+
+    def __init__(self):
+        self._project_ids = {}  # map of project_ids to their placeholder names
+        self._queries = []  # list of (key, promise, scalar subquery)
+
+    @classmethod
+    def current(cls):
+        if not hasattr(fg, 'node_counter'):
+            fg.node_counter = cls()
+        return fg.node_counter
+
+    def add_count(self, cls, args):
+        # escape non-trivial cases
+        if cls != psqlgraph.Node and not hasattr(cls, 'project_id') or \
+                cls.label == 'project' or list(args.keys()) != ['project_id']:
+            return
+
+        # extract project_id and guarantee permission
+        project_id = args['project_id']
+        if isinstance(project_id, (list, tuple)) and len(project_id) == 1:
+            project_id = project_id[0]
+        if not isinstance(project_id, (str, unicode)):
+            return
+        if project_id not in fg.read_access_projects:
+            return 0
+
+        # group project_id and name them
+        project_id_name = self._project_ids.get(project_id, None)
+        if project_id_name is None:
+            project_id_name = 'p_%s' % len(self._project_ids)
+            self._project_ids[project_id] = project_id_name
+
+        # prepare the subquery and promise
+        key = 'c_%s' % len(self._queries)
+        p = Promise()
+        self._queries.append((
+            key,
+            p,
+            self.query_tpl % (cls.__tablename__, project_id_name, key),
+        ))
+        return p
+
+    def run(self, session):
+        if not self._queries:
+            return
+
+        sql = 'SELECT %s;' % ', '.join(map(lambda t: t[2], self._queries))
```

Comment:
I think a list comprehension here is a little easier to read:
```suggestion
        sql = 'SELECT %s;' % ', '.join(count for _, _, count  in self._queries)
```